File: javascript/trees/same-tree.js
// Given the roots of two binary trees p and q, write a function to check if they are the same or not.

// Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.

class Node {
    constructor(val) {
        this.val = val
        this.left = null
        this.right = null
    }
}
// initilize nodes
const a = new Node(1) 
const b = new Node(2)
const c = new Node(3)
const d = new Node(1)
const e = new Node(2)
const f = new Node(3)


a.left = b
a.right = c
d.left = e
d.right = f


function sameTree(p, q) {
    // return true if both trees are empty
    if (!p && !q) return true
    // return false if one tree is empty
    if (!p || !q) return false
    // check for the same values in nodes
    if (p.val !== q.val) return false
    // check for the left and right 
    return (sameTree(p.left, q.left) && sameTree(p.right, q.right))
}

// testing
console.log(sameTree(a, d))
// expected output: true

File: python/trees/same_tree.py
# Given the roots of two binary trees p and q, write a function to check if they are the same or not.

# Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.

class Node:
    def __init__(self, val, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right

# Initialize the nodes
a = Node(1)
b = Node(2)
c = Node(3)
d = Node(1)
e = Node(2)
f = Node(3)

# Connect the nodes to form a binary search tree
a.left = b
a.right = c
d.left = e
d.right = f


def isSameTree(p: Node, q: Node) -> bool:
    # return true if both trees are empty
    if not p and not q:
        return True
    # return false if one tree is empty
    if not p or not q:
        return False
    # check for the same values in nodes
    if p.val != q.val:
        return False
    # check for the left and right 
    return (isSameTree(p.left, q.left) and isSameTree(p.right, q.right))

# testing
print(isSameTree(a, d))
# expected output: True